### PR TITLE
WebRTC Data Channel

### PR DESCRIPTION
**Pull request by ctrlaltdavid**
Implements the WebRTC data channel.

This PR is targeted at the `webapp` feature branch, where Web app-related changes will be accumulated and tested before being periodically merged into master (and vice-versa).

It builds upon PR https://github.com/vircadia/vircadia/pull/1221 which needs to be merged first. The changes in this PR start in commit `6c37628`.

QA: ~~See the Vircadia-Web-SDK PR: https://github.com/vircadia/vircadia-web-sdk/pull/6~~ See the Vircadia-Web-SDK PR: https://github.com/vircadia/vircadia-web-sdk/pull/10

Merge sequence:
- [x] Vircadia PR: https://github.com/vircadia/vircadia/pull/1219
- [x] Vircadia PR: https://github.com/vircadia/vircadia/pull/1239
- [x] Vircadia PR: https://github.com/vircadia/vircadia/pull/1258
- [x] Vircadia-Web-SDK PR: https://github.com/vircadia/vircadia-web-sdk/pull/6
- [x] Vircadia PR: https://github.com/vircadia/vircadia/pull/1261
- [x] Update Vircadia-Web repo to use the latest Vircadia-Web-SDK. 
Vircadia-Web PR: https://github.com/vircadia/vircadia-web/pull/13
- [x] Update this PR to use the latest Vircadia-Web `master` branch.
- [ ] CR, QA.
- [ ] Merge this PR.

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Interface-PR1238-e42c269-_c28f4749.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-e42c269-_c28f4749.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Interface-PR1238-1623b25-_645dc265.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-1623b25-.tar_645dc265.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-1623b25-.tar_645dc265.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-1623b25-_645dc265.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-1623b25-_645dc265.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-1623b25-_645dc265.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Interface-PR1238-a8ba2c5-_2402a5e3.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-a8ba2c5-_2402a5e3.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-a8ba2c5-.tar_2402a5e3.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-a8ba2c5-_2402a5e3.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-a8ba2c5-.tar_2402a5e3.Z

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Quest-Alpha-PR1238-a8ba2c5_2402a5e3.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-a8ba2c5_2402a5e3.apk

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Interface-PR1238-9a68e96-_28c408de.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-9a68e96-_28c408de.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-9a68e96-.tar_28c408de.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-9a68e96-.tar_28c408de.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-9a68e96-_28c408de.sh

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-9a68e96_28c408de.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Quest-Alpha-PR1238-9a68e96_28c408de.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-9a68e96-_28c408de.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-d5155f7-.tar_303ff9a7.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-d5155f7-.tar_303ff9a7.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-d5155f7-_303ff9a7.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Interface-PR1238-d5155f7-_303ff9a7.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-d5155f7-_303ff9a7.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Quest-Alpha-PR1238-d5155f7_303ff9a7.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-d5155f7_303ff9a7.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-d5155f7-_303ff9a7.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Interface-PR1238-c49ac16-_6b188d88.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c49ac16-_6b188d88.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c49ac16-.tar_6b188d88.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c49ac16-.tar_6b188d88.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c49ac16-_6b188d88.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c49ac16-_6b188d88.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c49ac16_6b188d88.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Quest-Alpha-PR1238-c49ac16_6b188d88.apk

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Interface-PR1238-6201675-_a0a0e8db.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-6201675-.tar_a0a0e8db.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-6201675-.tar_a0a0e8db.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-6201675-_a0a0e8db.sh

File: libraries/networking/CMakeLists.txt [hunk 1]
```diff
@@ -11,7 +11,8 @@ endif ()
 
 if (WIN32)
     # we need ws2_32.lib on windows, but it's static so we don't bubble it up
-    target_link_libraries(${TARGET_NAME} ws2_32.lib)
+    # Libraries needed for WebRTC: security.log winmm.lib
```
Comment (daleglass) [inline libraries/networking/CMakeLists.txt:14]: A bit pedantic, but 'security.log' is a typo

File: libraries/networking/src/webrtc/WebRTCDataChannels.cpp [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+void WDCConnection::onDataChannelStateChanged() {
+    auto state = _dataChannel->state();
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "WDCConnection::dataChannelStateChanged() :" << (int)state
+        << DataChannelInterface::DataStateString(state);
+#endif
+    if (state == DataChannelInterface::kClosed) {
+        _dataChannel->Close();
+        _dataChannel = nullptr;
+        // WEBRTC FIXME: The following line causes the _peerConnectionFactory to fail.
```
Comment (daleglass) [inline libraries/networking/src/webrtc/WebRTCDataChannels.cpp:285]: What happens then? Do we have a leak here meanwhile?

File: libraries/networking/src/webrtc/WebRTCDataChannels.cpp [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+    }
+
+    _parent->emitDataMessage(_dataChannelID, byteArray);
+}
+
+bool WDCConnection::sendDataMessage(const DataBuffer& buffer) {
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "WDCConnection::sendDataMessage()";
+#endif
+    const int MAX_WEBRTC_BUFFER_SIZE = 16 * 1024 * 1024;  // 16MB
```
Comment (daleglass) [inline libraries/networking/src/webrtc/WebRTCDataChannels.cpp:315]: Constant probably should be somewhere else, perhaps at the top of the file?
Why 16M? Is this a protocol issue, or just a sane memory limit?

What kind of data will be sent? Textures may exceed 16M quite easily.

There should be some sort of log message here, too.

File: libraries/shared/src/shared/WebRTC.h [hunk 4]
```diff
@@ -17,15 +17,15 @@
...
+// WEBRTC_DATA_CHANNELS: WebRTC client-server connections in parallel with UDP.
 
 #if defined(Q_OS_MAC)
 #  define WEBRTC_AUDIO 1
 #  define WEBRTC_POSIX 1
 #  define WEBRTC_LEGACY 1
 #elif defined(Q_OS_WIN)
 #  define WEBRTC_AUDIO 1
-#  define WEBRTC_DATA_CHANNEL 1
+#  define WEBRTC_DATA_CHANNELS 1
```
Comment (daleglass) [inline libraries/shared/src/shared/WebRTC.h:28]: Perhaps the constants should be named something like `ENABLE_WEBRTC_DATA_CHANNELS` instead, for clarity?

**Review by daleglass — CHANGES_REQUESTED** [review]

File: libraries/networking/src/webrtc/WebRTCDataChannels.cpp (hunk 2 above)
Comment (ctrlaltdavid) [reply libraries/networking/src/webrtc/WebRTCDataChannels.cpp:285]: Fixed

File: libraries/networking/src/webrtc/WebRTCDataChannels.cpp (hunk 3 above)
Comment (ctrlaltdavid) [reply libraries/networking/src/webrtc/WebRTCDataChannels.cpp:315]: This is a WebRTC protocol limit. This is just how much data can be buffered to send. Ultimately, it's called via `Socket::writeDatagram()` - in PR1259 - and UDP sockets can similarly fail to write date. The Vircadia protocol handles this situation (dropping unreliable packets and resending reliable packets). Large data items are chopped up into multi-packet messages.

File: libraries/shared/src/shared/WebRTC.h (hunk 4 above)
Comment (ctrlaltdavid) [reply libraries/shared/src/shared/WebRTC.h:28]: I think the "topic" of the define ("WEBRTC") coming first makes it easier to distinguish these defines from other defines and group related ones together.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c0e20b4-.tar_7ecd9b6b.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c0e20b4-_7ecd9b6b.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c0e20b4-.tar_7ecd9b6b.Z

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Interface-PR1238-c0e20b4-_7ecd9b6b.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c0e20b4-_7ecd9b6b.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c0e20b4_7ecd9b6b.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-Quest-Alpha-PR1238-c0e20b4_7ecd9b6b.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1238/Vircadia-PR1238-c0e20b4-_7ecd9b6b.exe

**Comment by ctrlaltdavid** [comment]
Latest changes should be QA's using Web SDK PR https://github.com/vircadia/vircadia-web-sdk/pull/10

**Comment by digisomni** [comment]
Le Wild Kalila has tested this again with vircadia/vircadia-web-sdk PR #10 by running a basic test suite.

**Review by daleglass — APPROVED** [review]

**Pull request merged**